### ISSUE TITLE
[gpu-next] HDR metadata not forwarded in windowed mode

### ISSUE THREAD

**Issue by mitzsch** [mpv-player/mpv#10628]
### Important Information

Provide following Information:
- mpv version: mpv-dev-x86_64-v3-20220909-git-d433c5d
- Windows Version: Windows 10 21H2
- Source of the mpv binary: shinchiro winbuild repo
- Hardware: AMD RX580 / Nvidia RTX3050

### Reproduction steps

I recently purchased a hdfury arcana to debug HDMI on my PC and other devices. I wanted to know if and which HDR metadata values mpv passes through to the TV. With mpv being in fullscreen, HDR turned on in Windows, and the following command line 
`mpv.com HDR.file -vo=gpu-next -hwdec=d3d11va -target-colorspace-hint=yes -v -fs`
it works perfectly, as seen in the picture.
![IMG_20220908_170002-cropped](https://user-images.githubusercontent.com/43502191/189478279-5f1f2aeb-117a-42dc-b9a5-04006dab4833.jpg)

However when running mpv with the same command just without `-fs` and therefore in windowed mode, no metadata is passed through. The hdfury shows the default windows HDR values. (For nvidia all 0, for AMD 1499maxCLL/799maxFALL) 
This behavior is quite unfortunate, as any software that embeds mpv and is not able to run it as fullscreen is out of luck passing through actual metadata. Also for daily use, where mpv should just run as a separate window it is unfortunate. It would be awesome if it could be added, either out of the box or with a dedicated _allow command_.

### Expected behavior

Allow for metadata passthrough even in windowed mode. 

### Actual behavior

No metadata forwarded when running mpv in windowed mode.

### Sample files

Any HDR file with metadata present

**Comment by ZaquL** [comment]
Does it work properly in vulkan? What about vulkan in SDR windows mode?

**Comment by mitzsch** [comment]
Thanks for your answer! I tried it with Vulkan on the RTX 3050 system. (On the AMD system it would not work at all with Vulkan) With the RTX 3050 the TV switched to HDR but received no metadata at all. (not in windowed nor in fullscreen mode) The hdfury showed 0 for all values. The only way I was able to get passthrough with Vulkan was with activated HDR under windows settings and even then it only worked in fullscreen. (note: information regarding colorspace were not transmitted with Vulkan - the hdfury showed unknown for target colorspace)

**Comment by mitzsch** [comment]
**update**

This seems to be a Microsoft / Windows problem. The sample code called `Direct3D 12 HDR sample` shows the same behavior, so it's not solely related to mpv.

https://github.com/microsoft/DirectX-Graphics-Samples/issues/797

**Comment by ZaquL** [comment]
If metadata will be forwarded in windowed mode, it will change the colors of Desktop wallpaper...

**Comment by mitzsch** [comment]
Valid point but in case where other programs are on the same screen (or the wallpaper), the windows compositor should handle the mapping.

**Comment by ZaquL** [comment]
It already does. What it does not do: handle what the TV will use as its
metadata algorithm. It is different for all TVs.

**Comment by ZaquL** [comment]
Hi! Can you check whether exclusive fullscreen works too:

Disable in mpv.exe propertities "fullscreen optimisation".

Then use this (it does not matter it is .com) **and press F:** mpv.com --d3d11-exclusive-fs=yes --target-colorspace-hint=yes --hwdec=d3d11va --gpu-api=d3d11 -vo=gpu-next HDRmeta.mp4

That only works in --gpu-api=d3d11. You can see that F is now not instantanious, because usually display switches. You can also see 

>[vo/gpu-next/d3d11] Enabling full-screen exclusive mode while entering fullscreen

**Comment by mitzsch** [comment]
Sure! 
I did some extensive testing with different settings on both the Nvidia RTX3050 and the AMD RX580 machine. (Both running Windows 10 21H2)

Here are the results:

AMD

- fullscreen optimization disabled, --d3d11-exclusive-fs=no
=> Passthrough only in fullscreen
=> Metadata not flushed when quitting in windowed mode
=> Metadata correctly flushed when quitting (or file ends) in fullscreen mode

- fullscreen optimization disabled, --d3d11-exclusive-fs=yes
=> when in fullscreen, HDR passthrough is completely broken, the TV even switches out of HDR, Image looks totally wrong

- fullscreen optimization enabled, --d3d11-exclusive-fs=no
=> Passthrough only in fullscreen
=> Metadata not flushed when quitting in windowed mode
=> Metadata correctly flushed when quitting (or file ends) in fullscreen mode => also when using --idle=yes

- fullscreen optimization enabled, --d3d11-exclusive-fs=yes
=> Passthrough only in fullscreen
=> Metadata correctly flushed when quitting in fullscreen or windowed mode (or file ends) => also when using --idle=yes



Nvidia

- fullscreen optimization disabled, --d3d11-exclusive-fs=no
=> Passthrough only in fullscreen
=> Metadata not flushed when quitting in windowed mode
=> When exiting fullscreen colorspace metadata is reset (brightness values are still in the hdmi stream)
=> Metadata correctly flushed when quitting (or file ends) in fullscreen mode

- fullscreen optimization disabled, --d3d11-exclusive-fs=yes
=> when in fullscreen, HDR passthrough is completely broken, the TV even switches out of HDR, Image looks totally wrong

- fullscreen optimization enabled, --d3d11-exclusive-fs=no
=> Passthrough only in fullscreen
=> Metadata not flushed when quitting in windowed or fullscreen mode


- fullscreen optimization enabled, --d3d11-exclusive-fs=yes
=> Passthrough only in fullscreen
=> Metadata correctly flushed when quitting in fullscreen or windowed mode (or file ends)
=> When using it with --idle=yes metadata is not flushed after file ends



Note to Nvidia:
Nvidia does some weird trickery. When HDR is enabled in the windows setting no brightness values are transmitted. When watching HDR files (especially the file used for testing here `"03. 900-4000nits-MaxCLL-4000-MDL-4000.mp4"` from [avforums](https://www.avsforum.com/threads/hdr10-test-patterns-set.2943380/)) the results seem fine even though no actual metadata is transmitted. (at least not visible for the hdfury arcana - I guess its true that there is no metadata in the stream - the tech support of hdfury told me that the arcana grabs everything on the HDMI link). On the AMD side, you can cleary see, when actual metadata is present. The  `"03. 900-4000nits-MaxCLL-4000-MDL-4000.mp4"`  file is clipped above 1500 nits (the default maxCLL value is 1499 for AMD) when watching with an AMD card and no metadata is present.

**Comment by ZaquL** [comment]
> * HDR passthrough is completely broken, the TV even switches out of HDR, Image looks totally wrong
> 
> *

You need to select 10 bit in nvidia control panel. Yes, I forgot to mention 8 bit will look horrible.

**Comment by mitzsch** [comment]
> You need to select 10 bit in nvidia control panel.

Well, 10bit is not a necessity for HDR. Nor on the AMD or the Nvidia side of things. I just tried it with 10bit on both systems and the result is the same. Using HDR with 8bit will introduce banding, but that's not what I meant to say. On both machines running `fullscreen optimization disabled, --d3d11-exclusive-fs=yes `, the TV will switch out of HDR back to SDR and therefore it's not surprising it will look off.

**Comment by ZaquL** [comment]
Yes, you are right, exclusive does not and dies not trigger hdr.

**Comment by Nevcairiel** [comment]
As far as I am aware, this is quite simply a Windows limitation. You don't need fullscreen exclusive mode, you just need a window that fully covers the entire screen perfectly. Only then will the Windows compositor let you take over full control of the output and use your HDR metadata, rather then its default values.

**Comment by ZaquL** [comment]
> What if multiple windows show HDR content, which metadata is favored?

The one on top, obviously, like with VRR. No, the problem is the inverse tonemapper for legacy sRGB content will not know how inverse tonemap cause there is no standard how to tonemap.

**Comment by Maxoverpower** [comment]
For anyone still struggling with this, you need to enable Windows' Automatic Color Management (ACM) to have automatic HDR instances display correctly in both windowed and fullscreen. It's currently found under Display -> Color profile (or Color Management).
On my end, new nvidia drivers don't work, which should be expected but I've heard one person report otherwise. Version 560 or 561 should work. Auto HDR with --target-colorspace-hint also doesn't work with d3d11, only vulkan, which happens to add almost a second to mpv's startup time and breaks AV1 hardware decode for my 3080. Your mileage may vary.

**Comment by netExtra** [comment]
> For anyone still struggling with this, you need to enable Windows' Automatic Color Management (ACM) to have automatic HDR instances display correctly in both windowed and fullscreen. It's currently found under Display -> Color profile (or Color Management). On my end, new nvidia drivers don't work, which should be expected but I've heard one person report otherwise. Version 560 or 561 should work. Auto HDR with --target-colorspace-hint also doesn't work with d3d11, only vulkan, which happens to add almost a second to mpv's startup time and breaks AV1 hardware decode for my 3080. Your mileage may vary.

On Windows 11, ACM is on by default in HDR. Not sure about Windows 10.

<img width="1152" height="168" alt="Image" src="https://github.com/user-attachments/assets/198b6c1f-99d1-486a-adca-37ac3710add8" />

**Comment by dyphire** [comment]
This seems to be a Windows 10 problem. Microsoft only has truly complete HDR support after Windows 11 24H2.
